libp2p/switch.nim

```diff
@@ -280,86 +291,96 @@ proc upgradeIncoming(s: Switch, conn: Connection) {.async, gcsafe.} =
+        raise newException(CatchableError, "Zombie connection encountered")
+
+      trace "Reusing existing connection", oid = $conn.oid,
+                                           direction = $conn.dir
+
+      return conn
+
+    trace "Dialing peer"
     for t in s.transports: # for each transport
       for a in addrs: # for each address
         if t.handles(a):   # check if it can dial it
-          trace "Dialing address", address = $a, peer = peerId
-          try:
-            conn = await t.dial(a)
-            # make sure to assign the peer to the connection
-            conn.peerInfo = PeerInfo.init(peerId, addrs)
-
-            conn.closeEvent.wait()
-              .addCallback do(udata: pointer):
-                asyncCheck s.triggerHooks(
-                  conn.peerInfo,
-                  Lifecycle.Disconnected)
-
-            asyncCheck s.triggerHooks(conn.peerInfo, Lifecycle.Connected)
-            libp2p_dialed_peers.inc()
-          except CancelledError as exc:
-            trace "dialing canceled", exc = exc.msg, peer = peerId
-            raise exc
-          except CatchableError as exc:
-            trace "dialing failed", exc = exc.msg, peer = peerId
-            libp2p_failed_dials.inc()
-            continue
-
-          try:
-            let uconn = await s.upgradeOutgoing(conn)
-            s.connManager.storeOutgoing(uconn)
-            asyncCheck s.triggerHooks(uconn.peerInfo, Lifecycle.Upgraded)
-            conn = uconn
-            trace "dial successful", oid = $conn.oid, peer = $conn.peerInfo
-          except CatchableError as exc:
-            if not(isNil(conn)):
-              await conn.close()
-
-            trace "Unable to establish outgoing link", exc = exc.msg, peer = peerId
-            raise exc
-
-          if isNil(conn):
-            libp2p_failed_upgrade.inc()
-            continue
+          trace "Dialing address", address = $a
+          let dialed = try:
+              await t.dial(a)
+            except CancelledError as exc:
+              trace "dialing canceled", exc = exc.msg
+              raise exc
+            except CatchableError as exc:
+              trace "dialing failed", exc = exc.msg
+              libp2p_failed_dials.inc()
+              continue # Try the next address
+
+          # make sure to assign the peer to the connection
+          dialed.peerInfo = PeerInfo.init(peerId, addrs)
+
+          libp2p_dialed_peers.inc()
+
+          let upgraded = try:
+              await s.upgradeOutgoing(dialed)
+            except CatchableError as exc:
+              # If we failed to establish the connection through one transport,
+              # we won't succeeed through another - no use in trying again
+              await dialed.close()
+              debug "upgrade failed", exc = exc.msg
+              if exc isnot CancelledError:
+                libp2p_failed_upgrade.inc()
+              raise exc
+
+          doAssert not isNil(upgraded), "checked in upgradeOutgoing"
```

Comment:
I've no idea what this message mean, if it dies here I've no hints to work with.

Reply:
well, the code randomly uses nil and exceptions when a function fails to perform its task, so this is a reminder that nil was checked in upgradeoutgoing and a second check is not needed here - this is more a reminder that if upgradeoutgoing changes, this code needs to change also